### PR TITLE
Allowing extensions to support `nodeByUri`

### PR DESCRIPTION
**Pull request by renatonascalves**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This adds two hooks allowing extensions to resolve their nodes when using `nodeByUri`.

Does this close any currently open issues?
------------------------------------------
no!

Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------

Screen Shot 2021-05-09 at 19 58 19

Any other comments?
-------------------
I also did a minor cleanup while I was browsing around the codebase.

Just for reference, here is where I'm using it: https://github.com/wp-graphql/wp-graphql-buddypress/issues/77

Where has this been tested?
---------------------------
**Operating System:** MacOS Big Sur 11.3

**WordPress Version:** Version 5.7.1

### CODE REVIEW

File: src/Type/InterfaceType/UniformResourceIdentifiable.php [hunk 1]
```diff
@@ -57,8 +56,14 @@ public static function register_type( TypeRegistry $type_registry ) {
...
-					return $type;
-
+					/**
+					 * Filter the UniformResourceIdentifiable node type.
+					 *
+					 * @param null|string $type The type definition the node should resolve to.
+					 * @param object $node The $node that is being resolved.
+					 * @param TypeRegistry $type_registry The type registry.
+					 */
+					return apply_filters( 'graphql_resolve_uniform_resource_identifiable_node_type', $type, $node, $type_registry );
```
Comment (jasonbahl) [inline src/Type/InterfaceType/UniformResourceIdentifiable.php:66]: @renatonascalves can you explain this filter a bit more? 

Since we're registering an InterfaceType here, this already gets passed through this filter here: https://github.com/wp-graphql/wp-graphql/blob/develop/src/Type/WPInterfaceType.php#L58

I'd prefer to keep filters more central like this, vs introducing new filters for every Type that's registered.

File: src/Data/NodeResolver.php [hunk 2]
```diff
@@ -45,6 +45,24 @@ public function __construct( AppContext $context ) {
...
+		 *
+		 * This is to be used in extensions to resolve their own nodes which might not use
+		 * WordPress permalink structure.
+		 *
+		 * @param null $node The node, defaults to nothing.
+		 * @param string $uri The uri being searched.
+		 * @param AppContext $content The app context.
+		 * @param WP $wp WP object.
+		 */
+		$node = apply_filters( 'graphql_pre_resolve_uri', null, $uri, $this->context, $this->wp );
```
Comment (jasonbahl) [inline src/Data/NodeResolver.php:60]: Nice. I think this is indeed a good addition!

File: src/Type/Union/PostObjectUnion.php [hunk 3]
```diff
@@ -41,7 +40,7 @@ public static function register_type( TypeRegistry $type_registry ) {
 	 *
 	 * @return array
 	 */
-	public static function get_possible_types() {
+	public static function get_possible_types(): array {
```
Comment (jasonbahl) [inline src/Type/Union/PostObjectUnion.php:43]: I think it could be good to start adding return types throughout, but let's do that in a separate PR?

File: src/Data/NodeResolver.php (hunk 2 above)
Comment (jasonbahl) [reply src/Data/NodeResolver.php:60]: Now, the trick is handling cases where the node to return is actually null. 🤔 

Like, what if we had some private data that we might resolve like so (psuedo code): 

```php
add_filter( 'graphql_pre_resolve_uri', function( $node, $uri ) {

   $node_in_db = some_function_to_get_node();
   
   if ( this_node_is_private( $node_in_db ) ) {
     return null;
   } 
   
   return $node;

});
```

In this example, we'd be resolving the data, determining there is no data, or the data isn't safe to show to the requesting user, so we return null. 

Now, the NodeResolver will have executed _that_ and will continue to execute the remaining logic instead of exiting early. 

🤔

File: src/Type/InterfaceType/UniformResourceIdentifiable.php (hunk 1 above)
Comment (renatonascalves) [reply src/Type/InterfaceType/UniformResourceIdentifiable.php:66]: Thanks for the link. Didn't know about that one. This is unnecessary. Will remove it. :)

File: src/Data/NodeResolver.php (hunk 2 above)
Comment (renatonascalves) [reply src/Data/NodeResolver.php:60]: @jasonbahl 

I put the hook in the top for a few reasons. Maybe very specific to my extension. 

- Currently, external URI is block.
- Homepage is blocked
- URLs return `null` if they are not match to any of the well known WP objects
- My extension URLs are customs, so there was no need to pass by the current `resolve_uri` logic first

In the scenario presented is okay to let the `resolve_uri` execution continues since the query is not type specific. It can match in a type and not in another. If you bail early, you might not get the node you need.

Comment (renatonascalves) [reply src/Data/NodeResolver.php:60]: And the node `Model` handles permissions. So if your node is private, it will return `null` and the execution can continue. Your private node won't show unless you have permissions to see it.

**Review by jasonbahl — APPROVED** [review]
